File: src/codes/w_04/complex_numbers_on_unit_circle_python_octave_cpp/python/complex_numbers_on_unit_circle_test.py
#!/usr/bin/python

#Onder Suvak (C) 2020

import numpy as np

rel_tol = 1e-06
abs_tol = 1e-10

no_pts = 100

fh_check_relative \
  = lambda computed, desired : \
      ( np.linalg.norm(computed - desired) \
          <= abs_tol + rel_tol * np.linalg.norm(desired) )

angles \
  = np.linspace(-np.pi, +np.pi, no_pts)
complex_points \
  = np.exp(1.j * angles)

angles_computed \
  = np.arctan2(
      np.imag(complex_points) ,
      np.real(complex_points) )

unity_norm_check_values \
  = np.abs( np.abs(complex_points) - 1. )
indices = np.where( unity_norm_check_values <= rel_tol )

flag_check_norm \
  = ( 0 == len( np.setdiff1d( 
                  range(len(complex_points)) , indices ) ) )
                  #np.arange(len(complex_points)) , indices ) ) )

print 'Check for unity norm of all values on the complex unit circle :'
if flag_check_norm:
  print '  YES'
else:
  print '  NO'

flag_check_angles \
  = fh_check_relative(angles_computed, angles);
  
print 'Check for compliance of each returned arg and original angle tuples :'
if flag_check_angles:
  print '  YES'
else:
  print '  NO'

File: src/codes/w_04/complex_numbers_on_unit_circle_python_octave_cpp/octave/complex_numbers_on_unit_circle_test.m
%  Onder Suvak (C) 2020

close all
clear variable
clear classes

rel_tol = 1e-06;
abs_tol = 1e-10;

no_pts = 100;

fh_check_relative ...
  = @(computed, desired) ...
      ( norm(computed - desired) ...
          <= abs_tol + rel_tol * norm(desired) );

angles ...
  = linspace(-pi, +pi, no_pts);
complex_points ...
  = exp(1j * angles);

angles_computed ...
  = atan2( ...
      imag(complex_points) , ...
      real(complex_points) );

unity_norm_check_values ...
  = abs( abs(complex_points) - 1 );
indices = find( unity_norm_check_values <= rel_tol );

flag_check_norm ...
  = isempty( setdiff( 1:numel(complex_points) , indices ) );

disp('Check for unity norm of all values on the complex unit circle :')
if flag_check_norm
  disp('  YES')
else
  disp('  NO')
end

flag_check_angles ...
  = fh_check_relative(angles_computed, angles);
  
disp('Check for compliance of each returned arg and original angle tuples :')
if flag_check_angles
  disp('  YES')
else
  disp('  NO')
end